### PR TITLE
Add SH1107_128x64 to the ssd1306 component

### PR DESCRIPTION
**Pull request by arunderwood**
# What does this implement/fix? 

Adding support for the [Adafruit FeatherWing OLED - 128x64 OLED](https://www.adafruit.com/product/4650), which uses the SH1107 driver and a resolution of 128x64.

PXL_20211230_224901027 (1)

### I2C vs SPI

Quoting a conversation had [on the docs PR](https://github.com/esphome/esphome-docs/pull/1765#discussion_r777023918)

>Based on how the ssd1306 component is written, theoretically the changes I have made to enable the use of a SH1107 over I2C also enabled the use over SPI, however the screen I'm testing with doesn't have the SPI interface populated, so the changes are untested.

>I'm fine with including this but I'll leave this note here for future people letting them know that I wasn't able to test SH1107 128x64 over SPI on real hardware.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** might fulfill https://github.com/esphome/feature-requests/issues/555, though I don't have the hardware to test it

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** [esphome/esphome-docs#<esphome-docs PR number goes here>](https://github.com/esphome/esphome-docs/pull/1765)

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
display:
  - platform: ssd1306_i2c
    model: 'SH1107_128x64'
    rotation: 90
    flip_x: false
    flip_y: false
    offset_y: -30
    lambda: |-
      it.line(0, 0, 128, 64);
      // Draw the outline of a rectangle with the top left at [5,5], a width of 10 and a height of 10
      it.rectangle(5, 5, 10, 10);
      it.circle(118, 54, 5);
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/ssd1306_base/__init__.py [hunk 1]
```diff
@@ -61,8 +62,8 @@ def _validate(value):
         cv.Optional(CONF_EXTERNAL_VCC): cv.boolean,
         cv.Optional(CONF_FLIP_X, default=True): cv.boolean,
         cv.Optional(CONF_FLIP_Y, default=True): cv.boolean,
-        cv.Optional(CONF_OFFSET_X, default=0): cv.int_range(min=0, max=15),
-        cv.Optional(CONF_OFFSET_Y, default=0): cv.int_range(min=0, max=15),
+        cv.Optional(CONF_OFFSET_X, default=0): cv.int_range(min=-32, max=32),
+        cv.Optional(CONF_OFFSET_Y, default=0): cv.int_range(min=-32, max=32),
```
Comment (jesserockz) [inline esphome/components/ssd1306_base/__init__.py:66]: Are these new offset ranges applicable to all screens, or could we add a validation to only allow -32 -> 32 for this new display model?

**Review by jesserockz — COMMENTED** [review]
LGTM, just one comment/question below

File: esphome/components/ssd1306_base/__init__.py (hunk 1 above)
Comment (arunderwood) [reply esphome/components/ssd1306_base/__init__.py:66]: Theoretically allowing the user to configure a greater offset range would increase the chance that a user could make any new SSD130X or SH110X based display work without having to modify the code.  

To be honest I'm not fully sure why a limit on an offset should be imposed at all.  I went back through the history to see if a reason was provided but it doesn't appear so https://github.com/esphome/esphome/pull/1902

**Review by jesserockz — APPROVED** [review]

**Review by oxan — APPROVED** [review]

**Comment by jenscski** [comment]
Hello

I also have one of the Adafruit OLED screens, and was thrilled when I found this support for it.
But I'm unable to make it work. I have, for my knowledge, installed the dev version of ESPHome, and I'm able to compile an example with the same code that's supplied in the config.yaml in the PR description.

But my screen is just white noise.

What I can see, is that the offset_y is set to -30, but in the logs I get this `Offset Y: 226`

```
[19:21:12][I][logger:214]: Log initialized
[19:21:12][I][app:029]: Running through setup()...
[19:21:12][I][i2c.arduino:159]: Performing I2C bus recovery
[19:21:12][C][ssd1306_i2c:010]: Setting up I2C SSD1306...
[19:21:12][I][app:062]: setup() finished successfully!
[19:21:12][I][app:102]: ESPHome version 2022.1.0-dev compiled on Jan  4 2022, 19:20:38
[19:21:12][C][logger:233]: Logger:
[19:21:12][C][logger:234]:   Level: DEBUG
[19:21:12][C][logger:235]:   Log Baud Rate: 115200
[19:21:12][C][logger:236]:   Hardware UART: UART0
[19:21:12][C][i2c.arduino:037]: I2C Bus:
[19:21:12][C][i2c.arduino:038]:   SDA Pin: GPIO23
[19:21:12][C][i2c.arduino:039]:   SCL Pin: GPIO22
[19:21:12][C][i2c.arduino:040]:   Frequency: 50000 Hz
[19:21:12][C][i2c.arduino:043]:   Recovery: bus successfully recovered
[19:21:12][I][i2c.arduino:053]: Results from i2c bus scan:
[19:21:12][I][i2c.arduino:059]: Found i2c device at address 0x3C
[19:21:12][I][i2c.arduino:059]: Found i2c device at address 0x58
[19:21:12][C][ssd1306_i2c:023]: I2C SSD1306
[19:21:12][C][ssd1306_i2c:023]:   Rotations: 90 °
[19:21:12][C][ssd1306_i2c:023]:   Dimensions: 128px x 64px
[19:21:12][C][ssd1306_i2c:024]:   Address: 0x3C
[19:21:12][C][ssd1306_i2c:025]:   Model: SH1107 128x64
[19:21:12][C][ssd1306_i2c:027]:   External VCC: NO
[19:21:12][C][ssd1306_i2c:028]:   Flip X: NO
[19:21:12][C][ssd1306_i2c:029]:   Flip Y: NO
[19:21:12][C][ssd1306_i2c:030]:   Offset X: 0
[19:21:12][C][ssd1306_i2c:031]:   Offset Y: 226
[19:21:12][C][ssd1306_i2c:032]:   Inverted Color: NO
[19:21:12][C][ssd1306_i2c:033]:   Update Interval: 1.0s
```

**Comment by jesserockz** [comment]
@jenscski Please create a new issue in https://github.com/esphome/issues

**Pull request merged**